src/util/util.ts

```diff
@@ -50,11 +50,20 @@ export function forEachReverse<T, U>(arr: readonly T[] | null | undefined, actio
     return undefined;
 }
 
-export function earliestDate(...dates: (Date | undefined)[]) {
-    return dates.reduce((res,d) => (res && d && d < res ? d : res) || d, undefined);
+function defaultCompare<T>(a: T, b: T) {
+    return a < b ? -1 : a > b ? 1 : 0;
 }
-export function latestDate(...dates: (Date | undefined)[]) {
-    return dates.reduce((res,d) => (res && d && d > res ? d : res) || d, undefined);
+
+export function min<T>(arr: readonly [T, ...(T | undefined)[]]): T;
+export function min<T>(arr: readonly T[], compare?: (a: T, b: T) => number): T | undefined;
+export function min<T>(arr: readonly T[], compare: (a: T, b: T) => number = defaultCompare) {
+    return arr.length === 0 ? undefined : arr.reduce((res, x) => compare(x, res) < 0 ? x : res);
+}
+
+export function max<T>(arr: readonly [T, ...(T | undefined)[]]): T;
+export function max<T>(arr: readonly T[], compare?: (a: T, b: T) => number): T | undefined;
+export function max<T>(arr: readonly T[], compare: (a: T, b: T) => number = defaultCompare) {
+    return arr.length === 0 ? undefined : arr.reduce((res, x) => compare(x, res) > 0 ? x : res);
```

Comment:
This looks good, but two questions:

1. Wouldn't it be more convenient to have the argument be a function (`getKey`) that returns a value to use for comparisons?
2. If not, then with this implementation it could be cute to implement one in terms of the other with the function negated...

Reply:
> Wouldn't it be more convenient to have the argument be a function (`getKey`) that returns a value to use for comparisons?

I think definitely for a `sort()` utility since it could potentially avoid repeated callbacks per element. For `min()` and `max()` I didn't see a strong indication either way, so I went with what seemed to be more common?
- https://github.com/microsoft/TypeScript/blob/33e7c729b9d6881a4f87af6b1c31e428d9e1b05d/src/compiler/core.ts#L1756
- https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List.html#v:minimumBy

> If not, then with this implementation it could be cute to implement one in terms of the other with the function negated...

Done :heavy_check_mark: